### PR TITLE
[WIP] Database V2

### LINKED ISSUE #33: Ensure cross-database consistency

### PR DESCRIPTION
**Pull request by cryptokat**
to resolve #33 

### Todos

- [x] unify all databases into one
- [x] group addBlock db operations into a batch
- [x] optimization: save database storage space for about 40% by delegate index (from ~14G to ~9G as of block 1365000)
- [ ] optimization: implement account index
- [ ] code cleanup

### CODE REVIEW

File: src/main/java/org/semux/consensus/SemuxBft.java [hunk 1]
```diff
@@ -995,12 +994,15 @@ protected void applyBlock(Block block) {
...
+            chain.addBlock(block);
+
+            // [8] commit state updates
             chain.getAccountState().commit();
             chain.getDelegateState().commit();
 
-            // [8] add block to chain
-            chain.addBlock(block);
+            // [9] commit pending blockchain updates
+            chain.commit();
```
Comment (orogvany) [inline src/main/java/org/semux/consensus/SemuxBft.java]: Should chain.commit also commit the delegate/account state?  Might ensure better consistency

File: src/main/java/org/semux/core/Block.java [hunk 2]
```diff
@@ -400,158 +397,13 @@ public long getTimestamp() {
...
-    }
-
     /**
      * Get block size in bytes
      *
      * @return block size in bytes
      */
     public int size() {
-        return toBytes().length;
+        return new BlockEncoderV1().encode(this).length;
```
Comment (orogvany) [inline src/main/java/org/semux/core/Block.java]: Will this always be v1 block?

**Comment by cryptokat** [comment]
Thanks for the review @orogvany. I will implement account index to see how much storage I can still save then fix those later.

File: src/main/resources/genesis/testnet.json [hunk 3]
```diff
@@ -7,27 +7,42 @@
...
         },
         {
             "address": "0x8e1ef4f810a84bea2fbf1e5f4afd8d0774ab0038",
+            "Abyte": "FIXME",
             "amount": 10000000000000000,
             "note": "development, marketing, promotion, bounties"
         },
         {
             "address": "0xc2ed00c8e05a2296ab3be4a0accb89b648bc71df",
+            "Abyte": "FIXME",
```
Comment (cryptokat) [inline src/main/resources/genesis/testnet.json]: Hi @semuxgo 
Could you publish the EDDSA public keys of testnet premine addresses? These are necessary to implement account index for saving storage space. Thanks

Comment (cryptokat) [reply src/main/resources/genesis/testnet.json]: ping @semuxgo

Comment (semuxgo) [reply src/main/resources/genesis/testnet.json]: Unfortunately, these testnet premined keys isn't properly saved. Is it possible to work around this?

Comment (cryptokat) [reply src/main/resources/genesis/testnet.json]: @semuxgo Not a problem. We can just reset the testnet with a new set of premine accounts later.

**Comment by semuxgo** [comment]
This is a nice feature to have. The majority work has been done by @cryptokat. I will wrap it up and finish the remaining tasks.

File: src/main/java/org/semux/core/Block.java (hunk 2 above)
Comment (semuxgo) [reply src/main/java/org/semux/core/Block.java]: We should propose a new block limit, considering that we're also enabling EVM.

**Pull request closed without merging**